app/src/main/kotlin/com/simplemobiletools/smsmessenger/helpers/MessagesReader.kt

```diff
@@ -0,0 +1,222 @@
+package com.simplemobiletools.smsmessenger.helpers
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.net.Uri
+import android.provider.Telephony.Mms
+import android.provider.Telephony.Sms
+import android.util.Base64
+import com.simplemobiletools.commons.extensions.*
+import com.simplemobiletools.commons.helpers.isQPlus
+import com.simplemobiletools.smsmessenger.models.MmsAddress
+import com.simplemobiletools.smsmessenger.models.MmsBackup
+import com.simplemobiletools.smsmessenger.models.MmsPart
+import com.simplemobiletools.smsmessenger.models.SmsBackup
+import java.io.IOException
+import java.io.InputStream
+
+class MessagesReader(private val context: Context) {
+    fun forEachSms(threadId: Long, block: (SmsBackup) -> Unit) {
+        val projection = arrayOf(
+            Sms.SUBSCRIPTION_ID,
+            Sms.ADDRESS,
+            Sms.BODY,
+            Sms.DATE,
+            Sms.DATE_SENT,
+            Sms.LOCKED,
+            Sms.PROTOCOL,
+            Sms.READ,
+            Sms.STATUS,
+            Sms.TYPE,
+            Sms.SERVICE_CENTER
+        )
+        val selection = "${Sms.THREAD_ID} = ?"
+        val selectionArgs = arrayOf(threadId.toString())
+        context.queryCursor(Sms.CONTENT_URI, projection, selection, selectionArgs) { cursor ->
+            val subscriptionId = cursor.getLongValue(Sms.SUBSCRIPTION_ID)
+            val address = cursor.getStringValue(Sms.ADDRESS)
+            val body = cursor.getStringValueOrNull(Sms.BODY)
```

Comment:
getStringValueOrNull is unresolved for me, have you maybe added it to Commons or somewhere?

Reply:
nevermind, added it into the latest commons

Reply:
Yes. I haven't pushed that yet. Sorry

Reply:
I pushed [here](https://github.com/SimpleMobileTools/Simple-Commons/pull/1145)